src/core/lib/surface/call_trace.cc

```diff
@@ -0,0 +1,114 @@
+    const grpc_channel_filter* filter) {
+  struct DerivedFilter : public grpc_channel_filter {
+    explicit DerivedFilter(const grpc_channel_filter* filter)
+        : grpc_channel_filter{
+              /* start_transport_stream_op_batch: */ grpc_call_next_op,
+              /* make_call_promise: */
+              [](grpc_channel_element* elem, CallArgs call_args,
+                 NextPromiseFactory next_promise_factory)
+                  -> ArenaPromise<ServerMetadataHandle> {
+                auto* source_filter =
+                    static_cast<const DerivedFilter*>(elem->filter)->filter;
+                gpr_log(
+                    GPR_DEBUG,
+                    "%sCreateCallPromise[%s]: client_initial_metadata=%s",
+                    Activity::current()->DebugTag().c_str(),
+                    source_filter->name,
+                    call_args.client_initial_metadata->DebugString().c_str());
+                return [source_filter, child = next_promise_factory(
+                                           std::move(call_args))]() mutable {
+                  gpr_log(GPR_DEBUG, "%sPollCallPromise[%s]: begin",
+                          Activity::current()->DebugTag().c_str(),
+                          source_filter->name);
+                  auto r = child();
+                  if (auto* p = absl::get_if<ServerMetadataHandle>(&r)) {
+                    gpr_log(GPR_DEBUG, "%sPollCallPromise[%s]: done: %s",
+                            Activity::current()->DebugTag().c_str(),
+                            source_filter->name, (*p)->DebugString().c_str());
+                  } else {
+                    gpr_log(GPR_DEBUG, "%sPollCallPromise[%s]: <<pending>",
+                            Activity::current()->DebugTag().c_str(),
+                            source_filter->name);
+                  }
+                  return r;
+                };
+              },
+              grpc_channel_next_op, /* sizeof_call_data: */ 0,
+              /* init_call_elem: */
+              [](grpc_call_element*, const grpc_call_element_args*) {
+                return absl::OkStatus();
+              },
+              grpc_call_stack_ignore_set_pollset_or_pollset_set,
+              /* destroy_call_elem: */
+              [](grpc_call_element*, const grpc_call_final_info*,
+                 grpc_closure*) {},
+              /* sizeof_channel_data: */ 0, /* init_channel_elem: */
+              [](grpc_channel_element*, grpc_channel_element_args*) {
+                return absl::OkStatus();
+              },
+              /* post_init_channel_elem: */
+              [](grpc_channel_stack*, grpc_channel_element*) {},
+              /* destroy_channel_elem: */ [](grpc_channel_element*) {},
+              grpc_channel_next_get_info, filter->name},
+          filter(filter) {}
+    const grpc_channel_filter* const filter;
+  };
+  struct Globals {
+    Mutex mu;
+    absl::flat_hash_map<const grpc_channel_filter*,
```

Comment:
Please add a lock annotation for this data member, just to be safe.

Reply:
done